test/emqtt_SUITE.erl

```diff
@@ -101,28 +139,120 @@ clean_retained(Topic) ->
 t_props(_) ->
     ok = emqtt_props:validate(#{'Payload-Format-Indicator' => 0}).
 
-t_connect(_) ->
-    {ok, C} = emqtt:start_link(),
-    {ok, _} = emqtt:connect(C),
+t_connect(Config) ->
+    ConnFun = ?config(conn_fun, Config),
+    Port = ?config(port, Config),
+    {ok, C} = emqtt:start_link([{port, Port}]),
+    {ok, _} = emqtt:ConnFun(C),
     ok= emqtt:disconnect(C),
-
-    {ok, C1} = emqtt:start_link([{clean_start, true}]),
-    {ok, _} = emqtt:connect(C1),
+    ct:pal("C is connected ~p", [C]),
+    {ok, C1} = emqtt:start_link([ {clean_start, true}
+                                , {port, Port}
+                                ]),
+    {ok, _} = emqtt:ConnFun(C1),
+    ct:pal("C1 is connected ~p", [C1]),
     ok= emqtt:disconnect(C1),
 
-    {ok, C2} = emqtt:start_link(#{clean_start => true}),
-    {ok, _} = emqtt:connect(C2),
+    {ok, C2} = emqtt:start_link(#{ clean_start => true
+                                 , port => Port
+                                 }
+                               ),
+    {ok, _} = emqtt:ConnFun(C2),
+    ct:pal("C2 is connected ~p", [C2]),
     ok= emqtt:disconnect(C2).
 
+t_reconnect_disabled(Config) ->
+    ConnFun = ?config(conn_fun, Config),
+    Port = ?config(port, Config),
+    process_flag(trap_exit, true),
+    {ok, C} = emqtt:start_link([{port, Port}]),
+    {ok, _} = emqtt:ConnFun(C),
+    MRef = erlang:monitor(process, C),
+    ok = emqtt_test_lib:stop_emqx(),
+    receive
+        {'DOWN', MRef, process, C, _Info} ->
+            receive
+                {'EXIT', C, {shutdown, tcp_closed}} when ConnFun =:= connect->
+                    ok;
+                {'EXIT', C, {shutdown, closed}} when ConnFun =:= quic_connect->
+                    ok
+            after 100 ->
+                    ct:fail(no_shutdown)
+            end
+    after 500 ->
+            ct:fail(conn_still_alive)
+    end.
+
+t_reconnect_enabled(Config) ->
+    ConnFun = ?config(conn_fun, Config),
+    Port = ?config(port, Config),
+    Topic = nth(1, ?TOPICS),
+    process_flag(trap_exit, true),
+    {ok, C} = emqtt:start_link([{port, Port},
+                                {reconnect, true},
+                                {connect_timeout, 1}]), % 1 sec
+    {ok, _} = emqtt:ConnFun(C),
+    MRef = erlang:monitor(process, C),
+    ok = emqtt_test_lib:stop_emqx(),
+    receive
+        {'DOWN', MRef, process, C, _Info} ->
+            ct:fail(conn_dead)
+    after 100 ->
+            timer:apply_after(5000, emqtt_test_lib, start_emqx, []),
+            {ok, _, [0]} = emqtt:subscribe(C, Topic),
+            {ok, C2} = emqtt:start_link([{port, Port}]),
+            {ok, _} = emqtt:ConnFun(C2),
```

Comment:
Just curiosity: if EMQX is dead and is only restarted after 5 s, does this block until that happens?  I thought this would raise a `{shutdown, econnrefused}` in such case.  :thinking:

Reply:
It is actually the L202 blocks because client `C` is in `reconnect` state and retrying which postpone all the event handling.